In [1]:
import pandas as pd

In [2]:
titanic_df = pd.read_csv("titanic_data.csv", delimiter = ',')

basic information

In [4]:
titanic_df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [5]:
titanic_df.shape

(891, 12)

In [7]:
titanic_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


The Titanic dataset has **891 rows and 12 columns** with passenger and survival information. Most columns are complete, but **Age, Cabin, and Embarked contain missing values**.


## Specific information

In [8]:
# Statistical Information
# mean mode median std min max correlation

In [9]:
titanic_df[["Age", "Fare"]]

,Age,Fare
0,22.0,7.2500
1,38.0,71.2833
2,26.0,7.9250
3,35.0,53.1000
4,35.0,8.0500
...,...,...
886,27.0,13.0000
887,19.0,30.0000
888,NaN,23.4500
889,26.0,30.0000


In [10]:
# mean
titanic_df[["Age", "Fare"]].mean()

Age     29.699118
Fare    32.204208
dtype: float64

In [11]:
# median
titanic_df[["Age", "Fare"]].median()

Age     28.0000
Fare    14.4542
dtype: float64

In [12]:
titanic_df[["Age", "Fare"]].describe()

,Age,Fare
count,714.000000,891.000000
mean,29.699118,32.204208
std,14.526497,49.693429
min,0.420000,0.000000
25%,20.125000,7.910400
50%,28.000000,14.454200
75%,38.000000,31.000000
max,80.000000,512.329200


a description for **each statistical factor**:

* **Count:** Age has **714 values**, meaning some ages are missing. Fare has **891 values**, so there are no missing fare values.
* **Mean:** The average passenger age is **29.7 years**, while the average fare is **32.20**.
* **Standard Deviation:** Age has a standard deviation of **14.53**, while Fare has **49.69**, showing that fares vary much more widely.
* **Minimum:** The youngest passenger was **0.42 years old**, and the minimum fare was **0**.
* **25% (Q1):** 25% of passengers were **20.13 years or younger**, and 25% paid **7.91 or less**.
* **50% (Median):** Half of the passengers were **28 years or younger**, and half paid **14.45 or less**.
* **75% (Q3):** 75% of passengers were **38 years or younger**, and 75% paid **31 or less**.
* **Maximum:** The oldest passenger was **80 years old**, while the highest fare was **512.33**.


In [19]:
# counts
titanic_df.value_counts(subset = "Survived", normalize = True, sort = True, ascending = True) #normalize =  percentage ko form ma dincha (0-1)
# sort = true indicates that highesh frequency stays at top, empty column true hudaina

Survived
1    0.383838
0    0.616162
Name: proportion, dtype: float64

In [14]:
titanic_df.Survived

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64

In [21]:
titanic_df.value_counts(subset = "Survived", normalize = True, dropna = True) #dropna True vayo bhane unique count hudaina

Survived
0    0.616162
1    0.383838
Name: proportion, dtype: float64

The survival data shows that **61.6% of the passengers did not survive**, while **38.4% survived**. This indicates that **more passengers lost their lives than survived** the Titanic disaster.


In [24]:
titanic_df.value_counts(["Sex", "Survived"], normalize = True)

Sex     Survived
male    0           0.525253
female  1           0.261504
male    1           0.122334
female  0           0.090909
Name: proportion, dtype: float64

The data shows that **52.5% of all passengers were males who did not survive**, while **12.2% were males who survived**. In comparison, **26.2% were females who survived**, while only **9.1% were females who did not survive**. This indicates that **female passengers had a much better survival outcome than male passengers**.


In [ ]:
# embarked from S, fare distribution

In [28]:
titanic_df[titanic_df.Embarked == "S"]["Fare"].describe()

count    644.000000
mean      27.079812
std       35.887993
min        0.000000
25%        8.050000
50%       13.000000
75%       27.900000
max      263.000000
Name: Fare, dtype: float64

In [34]:
# embarked from S and gender Male, Survival rate
titanic_df[
    (titanic_df.Embarked == "S") & (titanic_df["Sex"] == "male")
].value_counts("Survived")

Survived
0    364
1     77
Name: count, dtype: int64

## Crosstab - Compute a simple cross tabulation of two (or more) factors.

In [14]:
pd.crosstab(
    index = titanic_df.Sex,
    columns = titanic_df.Survived,
    margins = True,# gives the total number,default  =  All
    margins_name = "Total",
    colnames = ["Rate of survival"] #changing col name from surivied to rate of survival
)

Rate of survival,0,1,Total
Sex,,,
female,81,233,314
male,468,109,577
Total,549,342,891


In [16]:
# get unique values
titanic_df.Embarked.unique()

<StringArray>
['S', 'C', 'Q', nan]
Length: 4, dtype: str

In [19]:
titanic_df.Embarked.nunique(dropna = False) # gives the number of unique value including nan

4

In [21]:
# sum(), product(), variance(), cumsum(), mean(), quantile()
titanic_df.Age.quantile(q = 0.75)

np.float64(38.0)

In [23]:
# variance of Age
titanic_df.Age.var()

np.float64(211.01912474630802)

In [24]:
# mode of Embark
titanic_df.Embarked.mode()

0    S
Name: Embarked, dtype: str

### 0 - column wise -> vertical
### 1 - row wise -> horizontal

In [27]:
titanic_df.isna().sum(axis = 0) #gives empty value column wise

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [28]:
titanic_df.isna().sum(axis = 1) #gives empty value row wise

0      1
1      0
2      1
3      0
4      1
      ..
886    1
887    0
888    2
889    0
890    1
Length: 891, dtype: int64

In [29]:
# GroupBy
sample_df = titanic_df[["Sex", "Fare", "Age", "Survived", "Embarked"]]

In [30]:
sample_df.head()

,Sex,Fare,Age,Survived,Embarked
0,male,7.2500,22.0,0,S
1,female,71.2833,38.0,1,C
2,female,7.9250,26.0,1,S
3,female,53.1000,35.0,1,S
4,male,8.0500,35.0,0,S


In [31]:
group_by_gender = sample_df.groupby(by = "Sex")

In [33]:
group_by_gender.mean(numeric_only = True)

,Fare,Age,Survived
Sex,,,
female,44.479818,27.915709,0.742038
male,25.523893,30.726645,0.188908


In [34]:
group_by_gender.value_counts(["Survived"])

Sex     Survived
female  1           233
        0            81
male    0           468
        1           109
Name: count, dtype: int64

In [36]:
# multiple groupby
group_by_two =  titanic_df.groupby(by = ["Embarked", "Survived"])

In [38]:
group_by_two.mean(numeric_only = True)

PassengerId    Pclass        Age     SibSp     Parch  \
Embarked Survived                                                         
C        0          409.853333  2.200000  33.666667  0.253333  0.253333   
         1          473.989247  1.634409  28.973671  0.494624  0.451613   
Q        0          480.404255  2.936170  30.325000  0.510638  0.276596   
         1          319.966667  2.866667  22.500000  0.300000  0.000000   
S        0          449.868852  2.545667  30.203966  0.611241  0.348946   
         1          448.857143  1.967742  28.113184  0.493088  0.539171   

                        Fare  
Embarked Survived             
C        0         35.443335  
         1         79.720926  
Q        0         13.335904  
         1         13.182227  
S        0         20.743987  
         1         39.547081

### Correlation and Co-Variance

In [39]:
titanic_df[["Age", "Fare"]].cov()

,Age,Fare
Age,211.019125,73.849030
Fare,73.849030,2469.436846


In [42]:
titanic_df[["Age", "Fare"]].corr() # default =  pearson, scale =  -1,1

,Age,Fare
Age,1.000000,0.096067
Fare,0.096067,1.000000


0.096 indicates that there is no correlation 

### Sampling

In [47]:
sample_df_ = titanic_df.sample(n = 300, random_state = 999)

In [48]:
sample_df_.mean(numeric_only = True)

PassengerId    451.633333
Survived         0.366667
Pclass           2.353333
Age             29.880250
SibSp            0.530000
Parch            0.380000
Fare            31.413222
dtype: float64

In [49]:
sample_df_[["Age", "Fare"]].corr()

,Age,Fare
Age,1.00000,0.03441
Fare,0.03441,1.00000


Practice question

Using pd.crosstab(), create a table comparing:

Pclass as the rows
Survived as the columns
Add row and column totals
Rename the total label to "Total"
Rename the column-axis heading to "Survival Status"

In [5]:
pd.crosstab(
    index = titanic_df.Pclass,
    columns = titanic_df.Survived,
    normalize= True,
    margins=True,
    margins_name="Total",
    colnames=["Survival Status"]
           )

Survival Status,0,1,Total
Pclass,,,
1,0.089787,0.152637,0.242424
2,0.108866,0.097643,0.206510
3,0.417508,0.133558,0.551066
Total,0.616162,0.383838,1.000000


Practice question

Using pd.crosstab(), create a table that:

Uses Sex as the rows.
Uses Pclass as the columns.
Shows the proportion within each gender using normalize="index".
Renames the row heading to "Gender".
Renames the column heading to "Passenger Class".
Stores the result in sex_class_proportion.

In [7]:
sex_class_proportion = pd.crosstab(
    index= titanic_df.Sex,
    columns= titanic_df.Pclass,
    normalize= "index",
    rownames= ["Gender"],
    colnames= ["Passenger Class"]
)

In [8]:
sex_class_proportion

Passenger Class,1,2,3
Gender,,,
female,0.299363,0.242038,0.458599
male,0.211438,0.187175,0.601386
